packages/gensx-openai/src/index.tsx

```diff
@@ -1,18 +1,207 @@
+    });
+
+    // The function object has a type property that contains the OpenAI function definition
+    const tool = fn as {
+      type: "function";
+      function: ChatCompletionTool["function"];
+    };
+
+    return {
+      type: "function",
+      function: tool.function,
+    };
+  }
+}
+
+// Update the props types to use our new wrappers and extend from full OpenAI types
+type StreamingProps = Omit<
+  ChatCompletionCreateParamsStreaming,
+  "stream" | "tools" | "response_format"
+> & {
+  stream: true;
+  tools?: GSXTool<z.ZodObject<z.ZodRawShape>>[];
+  structuredOutput?: never;
+};
+
+type StructuredProps<T> = Omit<
+  ChatCompletionCreateParamsNonStreaming,
+  "stream" | "tools" | "response_format"
+> & {
+  stream?: false;
+  tools?: GSXTool<z.ZodObject<z.ZodRawShape>>[];
+  structuredOutput: GSXStructuredOutput<T>;
+};
+
+type StandardProps = Omit<
+  ChatCompletionCreateParamsNonStreaming,
+  "stream" | "tools" | "response_format"
+> & {
+  stream?: false;
+  tools?: GSXTool<z.ZodObject<z.ZodRawShape>>[];
+  structuredOutput?: never;
+};
+
+type GSXCompletionProps<T = unknown> =
+  | StreamingProps
+  | StructuredProps<T>
+  | StandardProps;
+
+type GSXCompletionReturn<P> = P extends StreamingProps
+  ? Stream<ChatCompletionChunk>
+  : P extends StructuredProps<infer T>
+    ? T
+    : ChatCompletionOutput;
+
+export const GSXCompletion = gsx.Component<
+  GSXCompletionProps,
+  GSXCompletionReturn<GSXCompletionProps>
+>("GSXCompletion", async (props) => {
+  const context = gsx.useContext(OpenAIContext);
+
+  if (!context.client) {
+    throw new Error(
+      "OpenAI client not found in context. Please wrap your component with OpenAIProvider.",
+    );
+  }
+
+  // Handle streaming case
+  if (props.stream) {
+    const { tools, ...rest } = props;
+    const openAITools = tools?.map((tool) => tool.toOpenAITool());
+
+    const response = await context.client.chat.completions.create({
+      ...rest,
+      stream: true,
+      tools: openAITools,
+    } as ChatCompletionCreateParamsStreaming);
+
+    return response as Stream<ChatCompletionChunk>;
+  }
+
+  // Handle structured output case
+  if (props.structuredOutput) {
+    const { tools, structuredOutput, ...rest } = props;
+    const openAITools = tools?.map((tool) => tool.toOpenAITool());
+
+    const completion = await context.client.chat.completions.create({
+      ...rest,
+      response_format: structuredOutput.toResponseFormat(),
+      tools: openAITools,
+    } as ChatCompletionCreateParamsNonStreaming);
+
+    const content = completion.choices[0].message.content;
+    if (!content) {
+      throw new Error("No content returned from OpenAI");
+    }
+
+    try {
+      const parsed = JSON.parse(content) as unknown;
+      const validated = structuredOutput.safeParse(parsed);
+      if (!validated.success) {
+        throw new Error(
+          `Invalid structured output: ${validated.error.message}`,
+        );
+      }
+      return validated.data;
```

Comment:
Would we want to use the structured output and chat completions components here?